File: dotnet/CardsProject/Domain/Marketplace/CardPriceHistory.cs
using CardsProject.Domain.Cards;
using System.ComponentModel.DataAnnotations.Schema;
using System.ComponentModel.DataAnnotations;

namespace CardsProject.Domain.Marketplace;

public class CardPriceHistory : IValidatableObject
{
    public int Id { get; set; }

    public DateOnly? PriceDate { get; set; } = null;
    public decimal AvgPrice { get; set; } = 0.00m;
    public decimal MinPrice { get; set; } = 0.00m;
    public decimal MaxPrice { get; set; } = 0.00m;
    public int Volume { get; set; } = 0;
    public bool Foil { get; set; } = false;

    public int? CardId { get; set; }
    [ForeignKey(nameof(CardId))]
    public Card? Card { get; set; }

    // Business operations

    public decimal PriceChangePercent(decimal previousAvg)
    {
        throw new NotImplementedException("price_change_percent not implemented");
    }

    public bool IsPriceSpike(int thresholdPercent)
    {
        throw new NotImplementedException("is_price_spike not implemented");
    }

    // ── Domain invariants (simple rules) ──────────────────────────────
    public IEnumerable<ValidationResult> Validate(ValidationContext validationContext)
    {
        if (!( (MinPrice <= AvgPrice && AvgPrice <= MaxPrice) ))
            yield return new ValidationResult("min_price <= avg_price <= max_price must hold", new[] { nameof(Id) });
        if (!( Volume >= 0 ))
            yield return new ValidationResult("Price history volume must not be negative", new[] { nameof(Id) });
        if (!( MinPrice >= 0m ))
            yield return new ValidationResult("Prices must not be negative", new[] { nameof(Id) });
    }
}

File: phoenix/lib/cards_project/marketplace/card_price_history.ex
defmodule CardsProject.Marketplace.CardPriceHistory do
  use Ecto.Schema
  import Ecto.Changeset

  schema "card_price_histories" do
    field :price_date, :date
    field :avg_price, :decimal
    field :min_price, :decimal
    field :max_price, :decimal
    field :volume, :integer
    field :foil, :boolean, default: false
    belongs_to :card, CardsProject.Cards.Card

    timestamps()
  end

  @doc false
  def changeset(record, attrs) do
    record
    |> cast(attrs, [:price_date, :avg_price, :min_price, :max_price, :volume, :foil, :card_id])
    |> validate_required([:price_date, :avg_price, :min_price, :max_price, :volume, :foil])
    |> validate_number(:volume, greater_than_or_equal_to: 0, message: "Price history volume must not be negative")
    |> validate_number(:min_price, greater_than_or_equal_to: 0, message: "Prices must not be negative")
  end

  # ── Business operations ────────────────────────────────────────────

  def price_change_percent(_record, _previous_avg) do
    # TODO: implement CardPriceHistory.price_change_percent
    {:error, :not_implemented}
  end

  def is_price_spike(_record, _threshold_percent) do
    # TODO: implement CardPriceHistory.is_price_spike
    {:error, :not_implemented}
  end
end
